File: quickSort.py
assignment_04-1-(2)
import time


class Arr:

    def __init__(self, word, mean):
        self.word = word
        self.mean = mean


def print_arr(word_arr):
    for item in word_arr:
        print(f"{item.word} : {item.mean}")


def adjust(word_arr, root, n):
    root_key = word_arr[root]
    root_index = root

    child = 2 * root + 1
    while child < n:
        if child + 1 < n and word_arr[child].word < word_arr[child + 1].word:
            child += 1
        if root_key.word >= word_arr[child].word:
            break
        word_arr[root_index] = word_arr[child]
        root_index = child
        child = 2 * root_index + 1
    word_arr[root_index] = root_key


def heap_sort(word_arr):
    n = len(word_arr)

    for i in range(n // 2 - 1, -1, -1):
        adjust(word_arr, i, n)

    for i in range(n - 1, 0, -1):
        word_arr[0], word_arr[i] = word_arr[i], word_arr[0]
        adjust(word_arr, 0, i)


def main():
    try:
        with open("randdict_utf8.TXT", "r", encoding="utf-8") as file:
            word_arr = []
            for line in file:
                parts = line.strip().split(":")
                if len(parts) >= 2:
                    word = parts[0].strip()
                    mean = parts[1].strip()
                    word_arr.append(Arr(word, mean))
                elif len(parts) == 1:
                    word = parts[0].strip()
                    mean = ""
                    word_arr.append(Arr(word, mean))
    except FileNotFoundError:
        print("No file")
        return

    start_time = time.time()
    heap_sort(word_arr)
    end_time = time.time()

    elapsed_time = end_time - start_time

    print("Sorted array:")
    print_arr(word_arr)

    print(f"Execution time: {elapsed_time} seconds")


if __name__ == "__main__":
    main()

File: quickSort.c
assignment_04-1-(2)_C
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX 48406
#define SWAP(x, y, t) ( (t)=(x), (x)=(y), (y)=(t) )

typedef struct arr{
  char *word;
  char *mean;
}arr;

void quick_sort(arr* wordArr, int left, int right);
int partition(arr* wordArr, int left, int right);

void print_Arr(arr* wordArr, int count){
  for(int i=0; i<count; i++){
    printf("%s : %s\n", wordArr[i].word, wordArr[i].mean);
  }
}

char *trim(char *str) {
  char *end;

  while (isspace((unsigned char)*str))
    str++;

  if (*str == 0) 
    return str;

  end = str + strlen(str) - 1;
  while (end > str && isspace((unsigned char)*end))
    end--;

  *(end + 1) = 0;

  return str;
}

void quick_sort(arr* wordArr, int left, int right)
{
  if (left < right) {
    int q = partition(wordArr, left, right);
    quick_sort(wordArr, left, q - 1);
    quick_sort(wordArr, q + 1, right);
  }
}
int partition(arr* wordArr, int left, int right)
{
  arr pivot, temp;
  int low, high;
  
  low = left;
  high = right + 1;
  pivot = wordArr[left];
  
  do {
    do
      low++;
    while (low <= right && (strcmp(wordArr[low].word, pivot.word) < 0));
    do
      high--;
    while (high >= left && (strcmp(wordArr[high].word, pivot.word) > 0));
    if (low < high) SWAP(wordArr[low], wordArr[high], temp);
  } while (low < high);
  SWAP(wordArr[left], wordArr[high], temp);
  return high;
}

int main() {
  FILE *file = fopen("randdict_utf8.TXT", "r");
  if (file == NULL) {
    perror("No file");
    return 1;
  }

  arr* wordArr = (arr*)malloc(sizeof(arr) * 50000);
  int count = 0;

  char buffer[100];
  while (fgets(buffer, sizeof(buffer), file) != NULL) {
    buffer[strcspn(buffer, "\n")] = '\0';

    char *token = strtok(buffer, ":");
    if (token != NULL) {
      arr bufArr;
      bufArr.word = strdup(trim(token));
      token = strtok(NULL, ":");
      if (token != NULL) {
        bufArr.mean = strdup(trim(token));
      } else {
        bufArr.mean = strdup("");
      }
      wordArr[count++] = bufArr;
    }
  }
  
  clock_t start_time = clock();
  
  quick_sort(wordArr, 0, count - 1);

  clock_t end_time = clock();

  double elapsed_time = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;

  printf("sorted array :\n");
  print_Arr(wordArr, count);

  printf("Execution time: %f seconds\n", elapsed_time);

  for (int i = 0; i < count; i++) {
    free(wordArr[i].word);
    free(wordArr[i].mean);
  }
  free(wordArr);

  fclose(file);

  return 0;

}